File: backend/app/models/ingestion.py
"""
Pydantic models for ingestion status API responses.
"""

from datetime import date, datetime

from pydantic import BaseModel, Field


class GTFSImportProgress(BaseModel):
    """Live progress of the GTFS static feed import."""

    state: str = "idle"
    phase: str | None = None
    message: str | None = None
    percent: float | None = None
    rows_processed: int | None = None
    rows_total: int | None = None
    started_at: datetime | None = None
    updated_at: datetime | None = None
    finished_at: datetime | None = None
    error_type: str | None = None
    error_message: str | None = None


class GTFSFeedStatus(BaseModel):
    """Status of the GTFS static feed import."""

    feed_id: str | None = None
    feed_url: str | None = None
    downloaded_at: datetime | None = None
    feed_start_date: date | None = None
    feed_end_date: date | None = None
    stop_count: int = 0
    route_count: int = 0
    trip_count: int = 0
    is_expired: bool = False
    import_progress: GTFSImportProgress = Field(default_factory=GTFSImportProgress)


class GTFSRTHarvesterStatus(BaseModel):
    """Status of the GTFS-RT realtime harvester."""

    is_running: bool = False
    last_harvest_at: datetime | None = None
    stations_updated_last_harvest: int = 0
    total_stats_records: int = 0


class IngestionStatus(BaseModel):
    """Combined ingestion status for GTFS static and realtime data."""

    gtfs_feed: GTFSFeedStatus
    gtfs_rt_harvester: GTFSRTHarvesterStatus

File: frontend/src/types/ingestion.ts
/**
 * Ingestion status types
 * Types for GTFS static feed and realtime harvester status
 */

export type GTFSImportProgressState = 'idle' | 'running' | 'succeeded' | 'failed'

export interface GTFSImportProgress {
  state: GTFSImportProgressState
  phase: string | null
  message: string | null
  percent: number | null
  rows_processed: number | null
  rows_total: number | null
  started_at: string | null
  updated_at: string | null
  finished_at: string | null
  error_type: string | null
  error_message: string | null
}

export interface GTFSFeedStatus {
  feed_id: string | null
  feed_url: string | null
  downloaded_at: string | null
  feed_start_date: string | null
  feed_end_date: string | null
  stop_count: number
  route_count: number
  trip_count: number
  is_expired: boolean
  import_progress: GTFSImportProgress
}

export interface GTFSRTHarvesterStatus {
  is_running: boolean
  last_harvest_at: string | null
  stations_updated_last_harvest: number
  total_stats_records: number
}

export interface IngestionStatus {
  gtfs_feed: GTFSFeedStatus
  gtfs_rt_harvester: GTFSRTHarvesterStatus
}
